### DISCUSSION TITLE
Is there a SPARQL query complexity analyzer similar to the ones available for GraphQL?

### DISCUSSION THREAD

**Discussion by Dimous** [eclipse-rdf4j/rdf4j#5172] [Q&A] [answered]
Hello!

I need to provide public access to SPARQL-enpoint.
So I plan to develop a request interceptor to bounce malicious queries.
Is there something like https://github.com/slicknode/graphql-query-complexity for SPARQL? Any (not necessarily for rdf4j or in java) implementation is welcome.

**Comment by kenwenzel** [comment, marked as answer]
RDF4J computes cost estimates during query optimization:
https://github.com/eclipse-rdf4j/rdf4j/blob/e25913a563a8de5bc4a88772dc074ab1f3c89c82/core/queryalgebra/evaluation/src/test/java/org/eclipse/rdf4j/query/algebra/evaluation/impl/QueryCostEstimatesTest.java#L25

You could try to interpret these values but I think it won't work for most of your cases as you would need the real result set cardinalities.

The best approach is to just limit the maximum query time by using `setMaxQueryTime()`:
https://rdf4j.org/javadoc/5.0.0/org/eclipse/rdf4j/query/Query.html#setMaxQueryTime(int)

**Reply by Dimous** [reply]
Thanks! I will look into it.
